kratos/utilities/exact_mortar_segmentation_utility.cpp

```diff
@@ -0,0 +1,1226 @@
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
+//
+//  Main authors:    Vicente Mataix Ferrandiz
+//
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "utilities/exact_mortar_segmentation_utility.h"
+
+namespace Kratos
+{
+template<>  
+bool ExactMortarIntegrationUtility<2,2, false>::GetExactIntegration(         
+    GeometryNodeType& OriginalSlaveGeometry,
+    const array_1d<double, 3>& SlaveNormal,
+    GeometryNodeType& OriginalMasterGeometry,
+    const array_1d<double, 3>& MasterNormal,
+    ConditionArrayListType& ConditionsPointsSlave
+    )
+{
+    // We take the geometry GP from the core 
+    const double tolerance = 1.0e-6; // std::numeric_limits<double>::epsilon();
+    
+    double total_weight = 0.0;
+    array_1d<double,2> auxiliar_coordinates = ZeroVector(2);
+    
+    // Declaring auxiliar values
+    PointType projected_gp_global;
+    GeometryNodeType::CoordinatesArrayType projected_gp_local;
+    
+    // First look if the edges of the slave are inside of the master, if not check if the opposite is true, if not then the element is not in contact
+    for (unsigned int i_slave = 0; i_slave < 2; i_slave++)
+    {
+        const array_1d<double, 3>& normal = OriginalSlaveGeometry[i_slave].GetValue(NORMAL);
+        
+        MortarUtilities::FastProjectDirection(OriginalMasterGeometry, OriginalSlaveGeometry[i_slave].Coordinates(), projected_gp_global, MasterNormal, -normal ); // The opposite direction
+        
+        const bool is_inside = OriginalMasterGeometry.IsInside( projected_gp_global.Coordinates( ), projected_gp_local, tolerance );
+        
+        if (is_inside == true) 
+        {
+            if (i_slave == 0)
+            {
+                auxiliar_coordinates[0] = - 1.0;
+            }
+            else if (i_slave == 1)
+            {
+                auxiliar_coordinates[1] =   1.0;
+            }
+        }
+    }
+    
+    if ((auxiliar_coordinates[0] == - 1.0 && auxiliar_coordinates[1] == 1.0) == true)
+    {
+        total_weight = 2.0;
+    }
+    else
+    {
+        std::vector<double> auxiliar_xi;
+        for (unsigned int i_master = 0; i_master < 2; i_master++)
+        {
+            projected_gp_local[0] = (i_master == 0) ? -1.0 : 1.0;
+            double delta_xi = (i_master == 0) ? 0.5 : -0.5;
+            const bool is_inside = MortarUtilities::ProjectIterativeLine2D(OriginalSlaveGeometry, OriginalMasterGeometry[i_master].Coordinates(), projected_gp_local, SlaveNormal, tolerance, delta_xi);
+            
+            if (is_inside == true)
+            {
+                auxiliar_xi.push_back(projected_gp_local[0]);
+            }
+        }
+        
+        if (auxiliar_xi.size() == 1 && ((auxiliar_coordinates[0] == - 1.0 || auxiliar_coordinates[1] == 1.0)))
+        {
+            if (std::abs(auxiliar_coordinates[0] + 1.0) < tolerance) // NOTE: Equivalent to == -1.0
```

Comment:
and this looks even more hacky...

Reply:
if you want to do this sort of things, please document it with A LOT of detail, otherwise no one will understand it (neither you) in 3 months from now